### ISSUE TITLE
[rllib] wrong target discount in DQN in case of truncated n_step

### ISSUE THREAD

**Issue by opherlieber** [ray-project/ray#4626]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **Ray installed from (source or binary)**: source
- **Ray version**: master branch
- **Python version**: 3.6
- **Exact command to reproduce**:

### Describe the problem

In DQN, when using 'truncate_episodes' batch_mode, it looks like the last <n_step-1> samples in each sample batch get a truncated n_step. In particular the last sample will get an n_step of 1 and it's target/NEXT_OBS will be the actual next observation. 
In the loss calculation though all samples get the full n_step discount (gamma**n_step) for the target value, even if they are truncated.

I don't know how critical if at all this is, I assume not so much for small n_step and/or large sample_batch_size, but for n_step of 10/20 the discount difference can be large.

**Comment by ericl** [comment]
Yeah, I think you are right, the target calculation should also take into account the truncated n_step. Though as you point out for small n the error is trivial.

cc @joneswong what do you think?

**Comment by joneswong** [comment]
yes, this is a problem. I am afraid that taking this into consideration by the TF computation graph is costly. How about post-processing `batch_size + n each` samples each time to generate `batch_size` ones for training, and then we continue interaction with the last `n` samples reserved.

**Comment by ericl** [comment]
Hm, that was the old strategy but it has bad behaviour around rollout boundaries.

Is it expensive to take the n into account? We could feed an array of the actual n step value used for each transition.

**Comment by joneswong** [comment]
I forgot the `assert not any(dones[:-1]), "Unexpected done in middle of trajectory"` assertion in the `_adjust_nstep` method. It seems that each postprocessed trajectory has only one done mask at the end. In this case, the last n-1 samples have `done=True` and thus `gamma ** n` would not introduce any bias.

**Comment by opherlieber** [comment]
> I forgot the `assert not any(dones[:-1]), "Unexpected done in middle of trajectory"` assertion in the `_adjust_nstep` method. It seems that each postprocessed trajectory has only one done mask at the end. In this case, the last n-1 samples have `done=True` and thus `gamma ** n` would not introduce any bias.

It only validates that the trajectories don't have a 'done' in the middle (i.e. only allows 'done; at the end, but doesn't require it), but the trajectories can still have no 'done' at all if they are from the middle of an episode (when using 'truncate_episodes'), this is usually the case for long episodes.

**Comment by joneswong** [comment]
my fault...so we cannot depend on that the last n-1 samples do not need the estimated target q values. Besides, as the correct discount factor changes according to the `done_masks` of the last n-1 samples, it is not viable to prepare a fixed array of discount factors. Thus, I think we'd better change `Sampler` to enable the old strategy.

**Comment by ericl** [comment]
We can compute a dynamic array of discounts and feed that in right? For example vtrace does this, though we feed in a constant array.

**Comment by joneswong** [comment]
> We can compute a dynamic array of discounts and feed that in right? For example vtrace does this, though we feed in a constant array.

I guess that [gamma**n, ..., gamma**n, gamma**(n-1), ..., gamma**1] is always viable, since we have correctly propagated the `done_mask` which enables transitions with `done_mask==True` to ignore the correct/incorrect discounts.

**Comment by joneswong** [comment]
@ericl Hi Eric, Am I wrong? or you will prepare a dynamic array?

> > We can compute a dynamic array of discounts and feed that in right? For example vtrace does this, though we feed in a constant array.
> 
> I guess that [gamma**n, ..., gamma**n, gamma**(n-1), ..., gamma**1] is always viable, since we have correctly propagated the `done_mask` which enables transitions with `done_mask==True` to ignore the correct/incorrect discounts.

**Comment by ericl** [comment]
That sounds like a reasonable solution to me, and also this idea of passing in the discounts separately per step seems pretty common in RL algorithms.

**Comment by joneswong** [comment]
@ericl  very sorry for my mistakes. I considered again. The problem is, for each (s, a, r, done, s') sampled from replay buffer, how could we know that the actual `n`. For those whose `done==True`, any `n` makes no difference. As for those whose `done==False`, we should have assigned appropriate `n` but currently we harshly give a fixed one. Without changing to the old strategy of sampling, we need the actual TD distance `n` to be in the training data. Otherwise, how could you feed the dynamic array of discount factors?

**Comment by ericl** [comment]
Could we store it in the replay buffer as well?

**Comment by joneswong** [comment]
@ericl definitely yes, but my concern is that most transitions are just corresponding to gamma \*\* n. If we add a field in replay buffer to store the discount factors, the developers may feel redundant and cumbersome.

**Comment by ericl** [comment]
Won't fix.
